### ISSUE TITLE
the state of DDPG

### ISSUE THREAD

**Issue by Maoquan232** [Tencent/PocketFlow#173]
As you mentioned that the states of DDPG are: 
###
    one-hot embedding of layer index
    weight tensor's shape
    number of parameters in the weight tensor
    number of quantization bits used by previous layers
    budget of quantization bits for remaining layers

which part of your code define the state vector and pass the state vector to the DDPG?
And why you chose these 5 factors as the state?

**Comment by jiaxiang-wu** [comment]
It depends on which learner you are using. For instance, for `WeightSparseLearner`, the state is computed in `learners/weight_sparsification/rl_helper.py`.
Basically, we follow the AMC paper (He et al., ECCV 2018) with a few modifications in defining the state vector.

**Comment by Maoquan232** [comment]
Thanks.
One more thing, can you explain what is the maskable_vars(list of maskable variables)?
In  learners/weight_sparsification/rl_helper.py 
I guess it is the layer?

**Comment by jiaxiang-wu** [comment]
Which line are you referring to (or, can you post the corresponding code block)?

**Comment by Maoquan232** [comment]
- [ ] class RLHelper(object):
- [ ]   """Reinforcement learning helper for the weight sparsification learner."""
- [ ] 
- [ ]   def __init__(self, sess, maskable_vars, skip_head_n_tail):
- [ ]     """Constructor function.
- [ ] 
- [ ]     Args:
- [ ]     * sess: TensorFlow session
- [ ]     * maskable_vars: list of maskable variables
- [ ]     * skip_head_n_tail: whether to skip the head & tail layers
- [ ]     """
- [ ] 
- [ ]     # obtain the shape & # of parameters of each maskable variable
- [ ]     nb_vars = len(maskable_vars)
- [ ]     var_shapes = []
- [ ]     self.prune_ratios = np.zeros(nb_vars)
- [ ]     self.nb_params_full = np.zeros(nb_vars)
- [ ]     for idx, var in enumerate(maskable_vars):
- [ ]       var_shape = sess.run(tf.shape(var))
- [ ]       assert var_shape.size in [2, 4], '# of variable dimensions is %d (invalid)' % var_shape.size
- [ ]       if var_shape.size == 2:
- [ ]         var_shape = np.hstack((np.ones(2), var_shape))
- [ ]       var_shapes += [var_shape]
- [ ]       self.nb_params_full[idx] = np.prod(var_shape)

**Comment by jiaxiang-wu** [comment]
For `WeightSparseLearner`, `maskable_vars` refers to:
1. convolutional kernels in convolutional layers (4-D tensor)
2. weighting matrix in dense / fully-connected layers (2-D tensor)

**Comment by Maoquan232** [comment]
hi, I noticed that you calculate the prune_ratios in one roll-out in the function __calc_rlout_actions(self) then pass it to __calc_optimal_prune_ratios(self) to calculate the reward.

  def __calc_rlout_actions(self):
    """Calculate actions within one roll-out.
    Returns:
    * prune_ratios: list of pruning ratios
    * states_n_actions: list of state vector and action pairs
    """

    self.agent.init_rlout()
    prune_ratios, states_n_actions = [], []
    for idx in range(len(self.vars_full['maskable'])):
      state = self.rl_helper.calc_state(idx)
      action = self.sess_train.run(self.agent.actions_noisy, feed_dict={self.agent.states: state})
      prune_ratio = self.rl_helper.cvt_action_to_prune_ratio(idx, action[0][0])
      prune_ratios += [prune_ratio]
      states_n_actions += [(state, action)]
      actor_loss, critic_loss, noise_std = self.agent.train()
    tf.logging.info('a-loss = %.2e | c-loss = %.2e | noise std. = %.2e'
                    % (actor_loss, critic_loss, noise_std))

    return prune_ratios, states_n_actions

def __calc_optimal_prune_ratios(self):
...
    for idx_rlout in range(FLAGS.ws_nb_rlouts):
      # compute actions & pruning ratios
      if is_primary_worker('global'):
        tf.logging.info('starting %d-th roll-out' % idx_rlout)
        prune_ratios, states_n_actions = self.__calc_rlout_actions()
        save_vals_to_file(prune_ratios, file_path_prune_ratios)
      if FLAGS.enbl_multi_gpu:
        self.mpi_comm.Barrier()
      prune_ratios = restore_vals_from_file(file_path_prune_ratios)

      # fine-tune the weight sparsified network to compute the reward
      reward = self.__calc_rlout_reward(prune_ratios)

Does it mean the agent can only get the reward after finishing pruning all the layers?  
Above codes are from  weight_sparsification/pr_optimizer.py

**Comment by jiaxiang-wu** [comment]
Yes, the DDPG agent can only get the reward after finishing pruning all the layers. The reward depends on the classification accuracy of the model with all layers pruned with certain pruning ratios.

**Comment by Maoquan232** [comment]
Thanks for reply
As the DDPG agent can only get the reward after finishing pruning all the layers，what is the reward before finishing？Set it to zero？

**Comment by jiaxiang-wu** [comment]
Reward is not needed during the roll-out. Here, a roll-out refers to the process of determining pruning ratios of all layers.
